usd/src/usd_parser/USDLinks.cc

```diff
@@ -117,22 +147,430 @@ void GetInertial(
+/// \param[in] _metersPerUnit meter per unit of the stage
+void ParseSphere(const pxr::UsdPrim &_prim,
+  sdf::Geometry &_geom,
+  const ignition::math::Vector3d &_scale,
+  double _metersPerUnit)
+{
+  double radius;
+  auto variant_sphere = pxr::UsdGeomSphere(_prim);
+  variant_sphere.GetRadiusAttr().Get(&radius);
+
+  sdf::Sphere s;
+  _geom.SetType(sdf::GeometryType::SPHERE);
+  s.SetRadius(radius * _metersPerUnit * _scale.X());
+  _geom.SetSphereShape(s);
+}
+
+//////////////////////////////////////////////////
+/// \brief Parse USD cylinder
+/// \param[in] _prim Reference to the USD prim
+/// \param[in] _geom sdf geom
+/// \param[in] _scale scale mesh
+/// \param[in] _metersPerUnit meter per unit of the stage
+void ParseCylinder(
+  const pxr::UsdPrim &_prim,
+  sdf::Geometry &_geom,
+  const ignition::math::Vector3d &_scale,
+  double _metersPerUnit)
+{
+  auto variant_cylinder = pxr::UsdGeomCylinder(_prim);
+  double radius;
+  double height;
+  variant_cylinder.GetRadiusAttr().Get(&radius);
+  variant_cylinder.GetHeightAttr().Get(&height);
+
+  sdf::Cylinder c;
+  _geom.SetType(sdf::GeometryType::CYLINDER);
+
+  c.SetRadius(radius * _metersPerUnit * _scale.X());
+  c.SetLength(height * _metersPerUnit * _scale.Y());
+
+  _geom.SetCylinderShape(c);
+}
+
+//////////////////////////////////////////////////
+UsdErrors ParseUSDLinks(
   const pxr::UsdPrim &_prim,
   const std::string &_nameLink,
   std::optional<sdf::Link> &_link,
-  const USDData &_usdData)
+  const USDData &_usdData,
+  ignition::math::Vector3d &_scale)
 {
+  UsdErrors errors;
   const std::string primNameStr = _prim.GetPath().GetName();
   const std::string primPathStr = pxr::TfStringify(_prim.GetPath());
   const std::string primType = _prim.GetPrimTypeInfo().GetTypeName().GetText();
 
+  const std::pair<std::string, std::shared_ptr<USDStage>> data =
+    _usdData.FindStage(primNameStr);
+
   // Is this a new link ?
   if (!_link)
   {
     _link = sdf::Link();
     _link->SetName(ignition::common::basename(_nameLink));
 
+    // USD define visual inside other visuals or links
+    // This loop allow to find the link for a specific visual
+    // For example
+    // This visual /ur10_long_suction/shoulder_link/cylinder
+    // correponds to the link /ur10_long_suction/wrist_3_link
+    // Then we can find the right transformations
+    std::string originalPrimName = pxr::TfStringify(_prim.GetPath());
+    size_t pos = std::string::npos;
+    std::string nameOfLink;
+    if ((pos = originalPrimName.find(_link->Name()))!= std::string::npos)
+    {
+      nameOfLink = originalPrimName.erase(
+        pos + _link->Name().length(),
+        originalPrimName.length() - (pos + _link->Name().length()));
+    }
+    pxr::UsdPrim tmpPrim = _prim;
+    if (!nameOfLink.empty())
+    {
+      while (tmpPrim)
+      {
+        if (pxr::TfStringify(tmpPrim.GetPath()) == nameOfLink)
+        {
+          break;
+        }
+        tmpPrim = tmpPrim.GetParent();
+      }
+    }
```

Comment:
Can we delete this code? Taking a closer look at it, it seems to define `tmpPrim`, which is a local variable that I don't see being used anywhere else. So, I don't see why this is needed.

Reply:
nope, I made this fix https://github.com/ignitionrobotics/sdformat/pull/897/commits/d8de8ba2e9bfb182ac3a7104cb300f5431372fdf

Nice catch